kong/tools/stream_api.lua

```diff
@@ -0,0 +1,114 @@
+
+local stream_api = {}
+
+local _endpoints = {}
+
+
+function stream_api.register(t)
+  for path, handler in pairs(t) do
+    if type(path) == "string" and type(handler) == "function" then
+      _endpoints[path] = handler
+    end
+  end
+end
+
+
+local stream_mt = {}
+stream_mt.__index = stream_mt
+
+function stream_api.serve()
+  local req = setmetatable({ socket = ngx.req.socket() }, stream_mt)
+
+  local path = req:get_path()
+  local handler = path and _endpoints[path]
+  if handler then
+    req:exit(handler(req))
+
+  else
+    req:exit(req:response("404 Not Found"))
+  end
+end
+
+
+function stream_mt:get_line()
+  if not self.req_line then
+    self.req_line = self.socket:receive("*l")
+  end
+
+  return self.req_line
+end
+
+function stream_mt:get_path()
+  if not self.path then
+    self.path = self:get_line():match("^%S+%s+(%S+)")
+  end
+
+  return self.path
+end
+
+function stream_mt:get_headers()
```

Comment:
Open question: I see the framework supports URI, headers and body at the same time. Do we need to support all these? Because it is for internal use, we can be a bit more opinionated and not support certain things that are not absolutely needed or are bit slower to process (like headers). It makes the addition of code less, and ensures everyone uses this system in a similar manner. (I see headers are not really tested in the cases below either)

Reply:
to avoid writing also the client side, i made it to be easy to use with resty.http; so it can't deviate toooooo much from HTTP.   in particular, that means reading headers.  

maybe we could ignore any `Content-Length` and just assume "close at the end"? that would allow not parsing into a table in simple cases.